<a href="https://colab.research.google.com/github/ItsMohamedAyman/NumPy/blob/main/NumPy_ufunc.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **1. Simple Arithmetic**

## Addition

In [2]:
import numpy as np

arr1 = np.array([10, 11, 12, 13, 14, 15])
arr2 = np.array([20, 21, 22, 23, 24, 25])

new_arr = np.add(arr1,arr2)
print(newarr)

[30 32 34 36 38 40]


## Subtraction

In [4]:
import numpy as np

arr1 = np.array([10, 20, 30, 40, 50, 60])
arr2 = np.array([20, 21, 22, 23, 24, 25])

newarr = np.subtract(arr1, arr2)

print(newarr)

[-10  -1   8  17  26  35]


## Multiplication

In [ ]:
import numpy as np

arr1 = np.array([10, 20, 30, 40, 50, 60])
arr2 = np.array([20, 21, 22, 23, 24, 25])

newarr = np.multiply(arr1, arr2)

print(newarr)

## Division

In [5]:
import numpy as np

arr1 = np.array([10, 20, 30, 40, 50, 60])
arr2 = np.array([3, 5, 10, 8, 2, 33])

newarr = np.divide(arr1, arr2)

print(newarr)

[ 3.33333333  4.          3.          5.         25.          1.81818182]


## Power

In [8]:
import numpy as np

arr1 = np.array([10, 20, 30, 40, 50, 60])
arr2 = np.array([3, 5, 6, 8, 2, 3])

newarr = np.power(arr1, arr2)

print(newarr)

[         1000       3200000     729000000 6553600000000          2500
        216000]


## Remainder

Both the mod() and the remainder() functions return the remainder of the values in the first array corresponding to the values in the second array, and return the results in a new array.

In [ ]:
import numpy as np

arr1 = np.array([10, 20, 30, 40, 50, 60])
arr2 = np.array([3, 7, 9, 8, 2, 33])

newarr = np.mod(arr1, arr2)

print(newarr)

You get the same result when using the remainder() function:

In [ ]:
import numpy as np

arr1 = np.array([10, 20, 30, 40, 50, 60])
arr2 = np.array([3, 7, 9, 8, 2, 33])

newarr = np.remainder(arr1, arr2)

print(newarr)

## Quotient and Mod

The divmod() function return both the quotient and the mod. The return value is two arrays, the first array contains the quotient and second array contains the mod.

In [9]:
import numpy as np

arr1 = np.array([10, 20, 30, 40, 50, 60])
arr2 = np.array([3, 7, 9, 8, 2, 33])

newarr = np.divmod(arr1, arr2)

print(newarr)

(array([ 3,  2,  3,  5, 25,  1]), array([ 1,  6,  3,  0,  0, 27]))


## Absolute Values

In [ ]:
import numpy as np

arr = np.array([-1, -2, 1, 2, 3, -4])

newarr = np.absolute(arr)

print(newarr)

# **2. Rounding Decimals**


## Truncation


Remove the decimals, and return the float number closest to zero. Use the trunc() and fix() functions.

In [10]:
import numpy as np

arr = np.trunc([-3.1666, 3.6667])

print(arr)

[-3.  3.]


Same example, using fix():

In [11]:
import numpy as np
arr = np.fix([-3.1666, 3.6667])
print(arr)

[-3.  3.]


## Rounding

The around() function increments preceding digit or decimal by 1 if >=5 else do nothing.

E.g. round off to 1 decimal point, 3.16666 is 3.2



In [20]:
import numpy as np

arr = np.around(3.1666, 2)

print(arr)

3.17


## Floor

The floor() function rounds off decimal to nearest lower integer.

E.g. floor of 3.166 is 3

In [21]:
import numpy as np

arr = np.floor([-3.1666, 3.6667])

print(arr)

[-4.  3.]


## Ceil

The ceil() function rounds off decimal to nearest upper integer.

E.g. ceil of 3.166 is 4

In [22]:
import numpy as np

arr = np.ceil([-3.1666, 3.6667])

print(arr)

[-3.  4.]


# **3. NumPy Logs**

## Log at Base 2

In [ ]:
import numpy as np

arr = np.arange(1, 10)

print(np.log2(arr))

## Log at Base 10

In [ ]:
import numpy as np

arr = np.arange(1, 10)

print(np.log10(arr))

## Natural Log, or Log at Base e

In [ ]:
import numpy as np

arr = np.arange(1, 10)

print(np.log(arr))

## Log at Any Base

In [24]:
from math import log
import numpy as np

nplog = np.frompyfunc(log, 2, 1)

print(nplog(100, 15))

1.7005483074552052


# **4. NumPy Summations**

## Summations

What is the difference between summation and addition?

Addition is done between two arguments whereas summation happens over n elements.

In [3]:
import numpy as np

arr1 = np.array([1,2,3,4])
arr2 = np.array([1,2,3,4])

newarr1 = np.add(arr1, arr2)
newarr2 = np.sum([arr1, arr2])


print(newarr1)
print(newarr2)

[2 4 6 8]
20


## Summation Over an Axis

If you specify axis=1, NumPy will sum the numbers in each array.

In [10]:
import numpy as np

arr1 = np.array([2, 4, 6])
arr2 = np.array([1, 2, 3])

newarr = np.sum([arr1, arr2], axis=1)

print(newarr)

[12  6]


🔧 ماذا تعني axis=1؟

axis=0: يجمع الصفوف (عموديًا).

axis=1: يجمع الأعمدة (أفقيًا).

## Cummulative Sum

Cummulative sum means partially adding the elements in array.

E.g. The partial sum of [1, 2, 3, 4] would be [1, 1+2, 1+2+3, 1+2+3+4] = [1, 3, 6, 10].

Perfom partial sum with the cumsum() function.

In [11]:
import numpy as np

arr = np.array([1, 2, 3])

newarr = np.cumsum(arr)

print(newarr)

[1 3 6]


# **5. NumPy Products**

## Products

In [12]:
import numpy as np

arr = np.array([1, 2, 3, 4])

x = np.prod(arr)

print(x)

24


2 × 3 × 4 = 24


In [13]:
import numpy as np

arr1 = np.array([1, 2, 3, 4])
arr2 = np.array([5, 6, 7, 8])

x = np.prod([arr1, arr2])

print(x)

40320


1 × 2 × 3 × 4 × 5 × 6 × 7 × 8


## Product Over an Axis

In [15]:
import numpy as np

arr1 = np.array([1, 2, 3, 4])
arr2 = np.array([5, 6, 7, 8])

newarr = np.prod([arr1, arr2], axis=1)
print(newarr)
newarr = np.prod([arr1, arr2], axis=0)
print(newarr)



[  24 1680]
[ 5 12 21 32]


## Cummulative Product

In [16]:
import numpy as np

arr = np.array([5, 6, 7, 8])

newarr = np.cumprod(arr)

print(newarr)

[   5   30  210 1680]


🧮 حساب النتيجة:
العنصر الأول:
5 → (مافيش قبله حاجة)

العنصر الثاني:
5 × 6 = 30

العنصر الثالث:
5 × 6 × 7 = 210

العنصر الرابع:
5 × 6 × 7 × 8 = 1680



# **6. NumPy Differences**

## **Differences**

In [17]:
import numpy as np

arr = np.array([10, 15, 25, 5])

newarr = np.diff(arr)

print(newarr)

[  5  10 -20]


نحسب الفروق بين كل عنصرين متتاليين:

15 - 10 = 5

25 - 15 = 10

5 - 25 = -20

In [18]:
import numpy as np

arr = np.array([10, 15, 25, 5])

newarr = np.diff(arr, n=2)

print(newarr)

[  5 -30]


نحسب الفروق الأولى:

15 - 10 = 5  
25 - 15 = 10  
5  - 25 = -20

🧮 خطوة 2 – diff الثانية (n=2):
نأخذ الناتج من الخطوة 1 ونحسب الفرق بين عناصره:

10 - 5  = 5  
-20 - 10 = -30

# **7. NumPy LCM**

##**Finding LCM (Lowest Common Multiple)**

In [19]:
import numpy as np

num1 = 4
num2 = 6

x = np.lcm(num1, num2)

print(x)

12


**الكود ده بيحسب الـ LCM (Least Common Multiple) أو "أصغر مضاعف مشترك" بين عددين باستخدام NumPy.**


📘 ما هو np.lcm()
الدالة np.lcm() ترجع أصغر عدد موجب يقبل القسمة على الرقمين بدون باقي.


🧮 مثال:
4 مضاعفاته: 4, 8, 12, 16, 20, ...

6 مضاعفاته: 6, 12, 18, 24, ...

أصغر رقم مشترك في المضاعفات = 12

## Finding LCM in Arrays

In [20]:
import numpy as np

arr = np.array([3, 6, 9])

x = np.lcm.reduce(arr)

print(x)

18


In [22]:
import numpy as np

arr = np.arange(1, 11)

x = np.lcm.reduce(arr)

print(arr)

print(x)

[ 1  2  3  4  5  6  7  8  9 10]
2520


# **8. NumPy GCD**

## Finding GCD

In [23]:
import numpy as np

num1 = 6
num2 = 9

x = np.gcd(num1, num2)

print(x)

3


Returns: 3 because that is the highest number both numbers can be divided by (6/3=2 and 9/3=3).

## Finding GCD in Arrays

In [24]:
import numpy as np

arr = np.array([20, 8, 32, 36, 16])

x = np.gcd.reduce(arr)

print(x)

4


#**9. NumPy Trigonometric**

## Trigonometric Functions

NumPy provides the ufuncs sin(), cos() and tan() that take values in radians and produce the corresponding sin, cos and tan values.

In [25]:
import numpy as np

x = np.sin(np.pi/2)

print(x)

1.0


In [26]:
import numpy as np

arr = np.array([np.pi/2, np.pi/3, np.pi/4, np.pi/5])

x = np.sin(arr)

print(x)

[1.         0.8660254  0.70710678 0.58778525]


## Convert Degrees Into Radians

In [27]:
import numpy as np

arr = np.array([90, 180, 270, 360])

x = np.deg2rad(arr)

print(x)

[1.57079633 3.14159265 4.71238898 6.28318531]


## Radians to Degrees

In [ ]:
import numpy as np

arr = np.array([np.pi/2, np.pi, 1.5*np.pi, 2*np.pi])

x = np.rad2deg(arr)

print(x)

## Finding Angles

Finding angles from values of sine, cos, tan. E.g. sin, cos and tan inverse (arcsin, arccos, arctan).

NumPy provides ufuncs arcsin(), arccos() and arctan() that produce radian values for corresponding sin, cos and tan values given.

In [ ]:
import numpy as np

x = np.arcsin(1.0)

print(x)

## Angles of Each Value in Arrays

In [ ]:
import numpy as np

arr = np.array([1, -1, 0.1])

x = np.arcsin(arr)

print(x)

## Hypotenues

In [28]:
import numpy as np

base = 3
perp = 4

x = np.hypot(base, perp)

print(x)

5.0


# **10. NumPy Hyperbolic**

## Hyperbolic Functions


NumPy provides the ufuncs sinh(), cosh() and tanh() that take values in radians and produce the corresponding sinh, cosh and tanh values..

In [ ]:
import numpy as np

x = np.sinh(np.pi/2)

print(x)

In [29]:
import numpy as np

arr = np.array([np.pi/2, np.pi/3, np.pi/4, np.pi/5])

x = np.cosh(arr)

print(x)

[2.50917848 1.60028686 1.32460909 1.20397209]


## Finding Angles


Finding angles from values of hyperbolic sine, cos, tan. E.g. sinh, cosh and tanh inverse (arcsinh, arccosh, arctanh).

Numpy provides ufuncs arcsinh(), arccosh() and arctanh() that produce radian values for corresponding sinh, cosh and tanh values given.

In [30]:
import numpy as np

x = np.arcsinh(1.0)

print(x)

0.881373587019543


In [ ]:
import numpy as np

arr = np.array([0.1, 0.2, 0.5])

x = np.arctanh(arr)

print(x)

# **11. NumPy Set Operations**

## What is a Set

A set in mathematics is a collection of unique elements.

Sets are used for operations involving frequent intersection, union and difference operations.

In [31]:
import numpy as np

arr = np.array([1, 1, 1, 2, 3, 4, 5, 5, 6, 7])

x = np.unique(arr)
print(x)

[1 2 3 4 5 6 7]


## Finding Union

In [32]:
import numpy as np

arr1 = np.array([1, 2, 3, 4])
arr2 = np.array([3, 4, 5, 6])

newarr = np.union1d(arr1,arr2)

print(newarr)

[1 2 3 4 5 6]


## Finding Intersection

In [34]:
import numpy as np

arr1 = np.array([1, 2, 3, 4])
arr2 = np.array([3, 4, 5, 6])

newarr = np.intersect1d(arr1, arr2, assume_unique=True)

print(newarr)

[3 4]


assume_unique=True يعني أنك تفترض أن المصفوفات لا تحتوي على تكرارات. هذا يجعل العملية أسرع.



## Finding Difference

In [35]:
import numpy as np

set1 = np.array([1, 2, 3, 4])
set2 = np.array([3, 4, 5, 6])

newarr = np.setdiff1d(set1, set2, assume_unique=True)

print(newarr)

[1 2]


## Finding Symmetric Difference


In [36]:
import numpy as np

set1 = np.array([1, 2, 3, 4])
set2 = np.array([3, 4, 5, 6])

newarr = np.setxor1d(set1, set2, assume_unique=True)

print(newarr)

[1 2 5 6]
